api/src/wfl/service/rawls.clj

```diff
@@ -37,14 +39,60 @@
    (create-snapshot-reference workspace snapshot-id name "")))
 
 (defn get-snapshot-reference
-  "Return the snapshot reference in fully-qualified Terra WORKSPACE with REFERENCE-ID."
+  "Return the snapshot reference in `workspace` with `reference-id`."
   [workspace reference-id]
-  (get-workspace-json workspace "snapshots" reference-id))
+  (get-workspace-json workspace snapshot-endpoint reference-id))
+
+(defn get-snapshot-references
+  "Lazily returns the snapshot references in `workspace`
+  with `limit` resources per page (default: 100)."
+  ([workspace limit]
+   (letfn
+    [(page [offset]
+       (let [{:keys [resources] :as _response}
+             (-> (workspace-api-url workspace snapshot-endpoint)
+                 (http/get {:headers      (auth/get-auth-header)
+                            :query-params {:offset offset
+                                           :limit  limit}})
+                 (util/response-body-json))
+             total    (+ offset (count resources))]
+         (lazy-cat resources (when (seq resources) (page total)))))]
+     (util/lazy-unchunk (page 0))))
+  ([workspace]
+   (get-snapshot-references workspace 100)))
+
+(defn create-or-get-snapshot-reference
+  "Return first snapshot reference for `snapshot-id` in `workspace`,
+  creating it as `name` with `description` if it does not yet exist."
+  ([workspace snapshot-id name description]
+   (try
+     (create-snapshot-reference workspace snapshot-id name description)
+     (catch ExceptionInfo cause
+       (let [{:keys [status] :as _data} (.getData cause)]
```

Comment:
`ex-data` is the clojure selector for this!

Reply:
Thank you thank you!